File: frontend/js/analytics.js
/* ================================================================
   ANALYTICS PAGE — Charts + Insights
   ================================================================ */

(async function initAnalytics() {
        const summaryGrid = document.getElementById("summary-grid");
        if (!summaryGrid) return;

        let salesChart = null;
        let categoryChart = null;

        // ---------- LOAD SUMMARY ----------
        async function loadSummary() {
            const res = await API.get("/api/reports/dashboard");
            if (!res.ok || !res.data.success) {
                showToast("Failed to load analytics", "error");
                return;
            }

            const { stats } = res.data;

            const cards = [
                { label: "Total Products", value: stats.total_products, cls: "" },
                { label: "Total Stock Units", value: stats.total_stock, cls: "" },
                { label: "Inventory Value", value: "₱" + Number(stats.inventory_value).toFixed(2), cls: "success" },
                { label: "Low Stock", value: stats.low_stock, cls: "warning" },
                { label: "Out of Stock", value: stats.out_of_stock, cls: "danger" },
                { label: "Categories", value: stats.total_categories, cls: "" },
            ];

            summaryGrid.innerHTML = cards.map((c) => `
            <div class="stat-card ${c.cls}">
                <div class="stat-label">${c.label}</div>
                <div class="stat-value">${c.value}</div>
            </div>
        `).join("");
        }

        // ---------- SALES TREND CHART ----------
        async function loadSalesChart() {
            const res = await API.get("/api/reports/transactions?limit=500");
            if (!res.ok || !res.data.success) return;

            const rows = res.data.report || [];
            const dailySales = {};

            rows.forEach((r) => {
                if (r.type !== "out" && r.type !== "sale") return;
                const d = new Date(r.created_at);
                const key = d.toISOString().slice(0, 10);
                dailySales[key] = (dailySales[key] || 0) + (r.quantity || 0);
            });

            // Last 30 days
            const labels = [];
            const values = [];
            for (let i = 29; i >= 0; i--) {
                const d = new Date();
                d.setDate(d.getDate() - i);
                const key = d.toISOString().slice(0, 10);
                labels.push(d.toLocaleDateString(undefined, { month: "short", day: "numeric" }));
                values.push(dailySales[key] || 0);
            }

            const ctx = document.getElementById("sales-chart");
            if (!ctx) return;

            if (salesChart) salesChart.destroy();

            salesChart = new Chart(ctx, {
                type: "line",
                data: {
                    labels,
                    datasets: [{
                        label: "Items Sold",
                        data: values,
                        borderColor: "#6366f1",
                        backgroundColor: "rgba(99, 102, 241, 0.12)",
                        fill: true,
                        tension: 0.35,
                        pointRadius: 2,
                        pointHoverRadius: 5,
                        borderWidth: 2,
                    }],
                },
                options: {
                    responsive: true,
                    maintainAspectRatio: false,
                    plugins: { legend: { display: false } },
                    scales: {
                        y: { beginAtZero: true, ticks: { precision: 0 } },
                        x: { grid: { display: false } },
                    },
                },
            });
        }

        // ---------- CATEGORY CHART ----------
        async function loadCategoryChart() {
            const res = await API.get("/api/categories");
            if (!res.ok || !res.data.success) return;

            const cats = res.data.categories || [];
            if (!cats.length) return;

            const labels = cats.map((c) => c.name);
            const values = cats.map((c) => c.product_count || 0);

            const palette = [
                "#6366f1", "#8b5cf6", "#ec4899", "#f59e0b",
                "#10b981", "#3b82f6", "#ef4444", "#14b8a6",
            ];

            const ctx = document.getElementById("category-chart");
            if (!ctx) return;

            if (categoryChart) categoryChart.destroy();

            categoryChart = new Chart(ctx, {
                type: "doughnut",
                data: {
                    labels,
                    datasets: [{
                        data: values,
                        backgroundColor: labels.map((_, i) => palette[i % palette.length]),
                        borderWidth: 2,
                        borderColor: "var(--panel)",
                    }],
                },
                options: {
                    responsive: true,
                    maintainAspectRatio: false,
                    plugins: {
                        legend: {
                            position: "bottom",
                            labels: { boxWidth: 10, padding: 10, font: { size: 11 } },
                        },
                    },
                    cutout: "60%",
                },
            });
        }

        // ---------- TOP PRODUCTS ----------
        async function loadTopProducts() {
            const list = document.getElementById("top-products-list");
            if (!list) return;

            const res = await API.get("/api/reports/transactions?limit=1000");
            if (!res.ok || !res.data.success) {
                list.innerHTML = `<p style="color:var(--muted);font-size:13px;">No data</p>`;
                return;
            }

            const rows = res.data.report || [];
            const counts = {};

            rows.forEach((r) => {
                if (r.type !== "out" && r.type !== "sale") return;
                const name = r.product || r.product_name || "Unknown";
                counts[name] = (counts[name] || 0) + (r.quantity || 0);
            });

            const sorted = Object.entries(counts)
                .sort((a, b) => b[1] - a[1])
                .slice(0, 5);

            if (!sorted.length) {
                list.innerHTML = `<p style="color:var(--muted);font-size:13px;">No sales data yet</p>`;
                return;
            }

            list.innerHTML = sorted.map(([name, qty], i) => `
            <div class="top-item">
                <span class="top-item-name">${i + 1}. ${escapeHtml(name)}</span>
                <span class="top-item-meta">${qty} sold</span>
            </div>
        `).join("");
        }

        // ---------- LOW STOCK ALERTS ----------
        async function loadLowStock() {
            const list = document.getElementById("low-stock-list");
            if (!list) return;

            const res = await API.get("/api/reports/low-stock");
            if (!res.ok || !res.data.success) {
                list.innerHTML = `<p style="color:var(--muted);font-size:13px;">No data</p>`;
                return;
            }

            const rows = res.data.report || [];
            if (!rows.length) {
                list.innerHTML = `<p style="color:var(--muted);font-size:13px;">All products are well-stocked ✅</p>`;
                return;
            }

            list.innerHTML = rows.slice(0, 8).map((r) => {
                        const out = (r.current_stock || 0) <= 0;
                        return `
                <div class="low-item ${out ? "out" : ""}">
                    <span>${escapeHtml(r.name || r.sku)}</span>
                    <span>${out ? "OUT OF STOCK" : `${r.current_stock} left (min ${r.min_stock_level})`}</span>
                </div>
            `;
        }).join("");
    }

    // ---------- REFRESH ----------
    async function loadAll() {
        await Promise.all([
            loadSummary(),
            loadSalesChart(),
            loadCategoryChart(),
            loadTopProducts(),
            loadLowStock(),
        ]);
    }

    const refreshBtn = document.getElementById("refresh-btn");
    if (refreshBtn) {
        refreshBtn.addEventListener("click", async () => {
            await loadAll();
            showToast("Analytics refreshed", "success");
        });
    }

    await loadAll();
})();

File: backend/routes/analytics.py
"""Analytics API routes - aggregated stats and trends."""
from flask import Blueprint, jsonify
from services.db import fetch_one, fetch_all
from security import login_required, role_required

analytics_bp = Blueprint("analytics", __name__)


@analytics_bp.get("/dashboard")
@login_required
def dashboard():
    """Return aggregated dashboard stats."""
    try:
        stats = fetch_one("""
            SELECT
                (SELECT COUNT(*) FROM products WHERE status = 'active') AS total_products,
                (SELECT COALESCE(SUM(current_stock), 0) FROM products WHERE status = 'active') AS total_stock,
                (SELECT COALESCE(SUM(current_stock * cost_price), 0) FROM products WHERE status = 'active') AS inventory_value,
                (SELECT COUNT(*) FROM products WHERE status = 'active' AND current_stock <= min_stock_level AND current_stock > 0) AS low_stock,
                (SELECT COUNT(*) FROM products WHERE status = 'active' AND current_stock <= 0) AS out_of_stock,
                (SELECT COUNT(*) FROM categories) AS total_categories,
                (SELECT COUNT(*) FROM suppliers) AS total_suppliers
        """) or {}

        return jsonify({"success": True, "stats": stats})
    except Exception as e:
        return jsonify({"success": False, "message": str(e)}), 500


@analytics_bp.get("/sales-trend")
@login_required
def sales_trend():
    """Return sales trend for last 30 days."""
    try:
        rows = fetch_all("""
            SELECT
                DATE(created_at) AS day,
                COALESCE(SUM(quantity), 0) AS total_qty
            FROM inventory_transactions
            WHERE type IN ('out', 'sale')
              AND created_at >= NOW() - INTERVAL '30 days'
            GROUP BY DATE(created_at)
            ORDER BY day ASC
        """)
        return jsonify({"success": True, "data": rows or []})
    except Exception as e:
        return jsonify({"success": False, "message": str(e)}), 500


@analytics_bp.get("/top-products")
@login_required
def top_products():
    """Return top-selling products (last 30 days)."""
    try:
        rows = fetch_all("""
            SELECT
                p.name,
                p.sku,
                COALESCE(SUM(t.quantity), 0) AS total_sold
            FROM inventory_transactions t
            JOIN products p ON p.id = t.product_id
            WHERE t.type IN ('out', 'sale')
              AND t.created_at >= NOW() - INTERVAL '30 days'
            GROUP BY p.id, p.name, p.sku
            ORDER BY total_sold DESC
            LIMIT 10
        """)
        return jsonify({"success": True, "data": rows or []})
    except Exception as e:
        return jsonify({"success": False, "message": str(e)}), 500


@analytics_bp.get("/low-stock")
@login_required
def low_stock():
    """Return products with low stock."""
    try:
        rows = fetch_all("""
            SELECT id, sku, name, current_stock, min_stock_level
            FROM products
            WHERE status = 'active'
              AND current_stock <= min_stock_level
            ORDER BY current_stock ASC
            LIMIT 50
        """)
        return jsonify({"success": True, "data": rows or []})
    except Exception as e:
        return jsonify({"success": False, "message": str(e)}), 500